designing-for-reusability.md

```diff
@@ -0,0 +1,158 @@
+package.json   manifest file
+```
+This structure is only here as a reference (we're not going to tell you how to
+write your apps). But it's a model we've found to work well for `choo` apps as
+it only works well for slim applications, and focuses on splitting things off.
+So if you haven't decided on a structure yet, this might be a nice one to try.
+
+## Effects
+Probably the premier candidate to modularize would be effects. Effects are
+generally the places where most complexity hides; calling the `choo` specific
+`send(name, data, cb)` function to issue commands to the rest of the
+application. Though very powerful, using `send()` all over the place will
+eventually cause for mess and giant fallout. Instead there's several patterns
+that might help:
+
+### send() at the borders only
+[ talk about callbacks and never passing `send()` directly ]
+
+### Rise of the SDKs
+[ talk about how SDKs can wrap APIs, making networking reusable ]
+
+## Views
+Views are passed three things: the current state of the world, the state of the
+world like it was in the last frame that was rendered and a `send()` method
+that can issue commands to update the state of the world. These things are the
+APIs `choo` exposes to its view components (aka components that make up views).
+
+As with all other parts of the framework, we should try and minimize the extent
+of interaction between the components we create and the framework's API. An
+approach that works particularly nice is to create elements using a function
+which accepts an object with callbacks on it that are propagated into the
+element. Whenever an action is triggered, it calls the appropriate callback
+passed in.
+
+Another important thing to keep in mind is that not all components need all
+data. Often time it's worthwhile to deconstruct `state` and only pass relevant
+fields into the components. This makes it easier to create testable components,
+and better defines the scope of an element.
+
+Say we have a form element. We want it to trigger two types of validation: when
+input occurs we want to validate it on the client. When submit is clicked we
+want it to submit the data and validate it on the server. To create a
+generalized component we'd do:
+```js
+// ./pages/main.js
+const html = require('choo/html')
+const Form = require('../components/form')
+
+const renderForm = Form(formInput, {
+  onInput: (data) => send('myFormModel:validate', data),
+  onSubmit: (data) => send('myFormModel:submit', data)
+})
+
+module.exports = function mainView (state, prev, send) {
+  const formInput = state.myFormModel.input
+
+  return html`
+    <main>
+      ${renderForm(formInput)}
+    </main>
+  `
+}
+```
+
+```js
+// ./components/form.js
+const getFormData = require('form-data')
+const html = require('choo/html')
+const assert = require('assert')
+
+module.exports = function formComponent (evMap) {
+  assert.equal(typeof evMap, 'object')
+
+  return function (input) {
```

Comment:
It's being thunked; to have it like have an "init" stage and a "render" stage; the API for the consumers looks kinda nice this way, and there's no way of accidentally redeclaring the objects inline which would make it super slow. Does that make sense? Should we explain that better?